lightning_uq_box/uq_methods/laplace_model.py

```diff
@@ -28,29 +27,44 @@
 # over training_step in lightning
 
 
-def tune_prior_precision(
-    model: Laplace, tune_precision_lr: float, n_epochs_tune_precision: int
+def tune_prior_precision_and_sigma(
+    model: Laplace,
+    tune_precision_lr: float,
+    n_epochs_tune_precision: int,
+    tune_prior_precision: bool,
+    tune_sigma_noise: bool,
 ):
-    """Tune the prior precision via Empirical Bayes.
+    """Tune the prior precision and sigma noise via Empirical Bayes.
 
     Args:
         model: laplace model
         tune_precision_lr: learning rate for tuning prior precision
         n_epochs_tune_precision: number of epochs to tune prior precision
+        tune_prior_precision: whether to tune prior precision
+        tune_sigma_noise: whether to tune sigma noise
     """
-    log_prior, log_sigma = (
-        torch.ones(1, requires_grad=True),
-        torch.ones(1, requires_grad=True),
-    )
-    hyper_optimizer = torch.optim.Adam([log_prior, log_sigma], lr=tune_precision_lr)
-    bar = trange(n_epochs_tune_precision)
-    # find out why this is so extremely slow?
-    for i in bar:
-        hyper_optimizer.zero_grad()
-        neg_marglik = -model.log_marginal_likelihood(log_prior.exp(), log_sigma.exp())
-        neg_marglik.backward()
-        hyper_optimizer.step()
-        bar.set_postfix(neg_marglik=f"{neg_marglik.detach().cpu().item()}")
+    with torch.inference_mode(False):
+        log_prior, log_sigma = (
+            torch.ones(1, requires_grad=True),
+            torch.ones(1, requires_grad=True),
+        )
+        optim_params = []
+        if tune_prior_precision is not None:
+            optim_params.append(log_prior)
+        if tune_sigma_noise is not None:
+            optim_params.append(log_sigma)
+
+        hyper_optimizer = torch.optim.Adam(optim_params, lr=tune_precision_lr)
+        bar = trange(n_epochs_tune_precision)
+        # find out why this is so extremely slow?
+        for i in bar:
+            hyper_optimizer.zero_grad()
+            neg_marglik = -model.log_marginal_likelihood(
+                log_prior.exp(), log_sigma.exp()
+            )
```

Comment:
I really don't mean to be a stickler here, but this one caught my eye, and since you asked for a review :) Looking at [the code for `log_marginal_likelihood`](https://github.com/aleximmer/Laplace/blob/7af7e30b3875657a47eb248c99769071521b7d65/laplace/baselaplace.py#L673), it has a documented but still unexpected side effect: 

```
By passing `prior_precision` or `sigma_noise`, the current value is
overwritten. This is useful for iterating on the log marginal likelihood.
```

meaning that it sets `model.prior_precision = prior_precision` and `model.sigma_noise = sigma_noise` in each call. So even if users say e.g. `tune_sigma_noise=False`, the code in `tune_prior_precision_and_sigma()` will overwrite that with `exp(1) = e`. If users do not set custom values `Laplace(..., sigma_noise=..., prior_precision=...)`, also the defaults (both 1.0) will be changed, as far as I can tell. An easy fix would be to use `model.sigma_noise` in the `log_marginal_likelihood()` call in place of `log_sigma.exp()` in that case, same for `prior_precision` of course.

All other changes look fine I think. Without having given the library a good shakedown still and if your tests pass, then I think I'm in no position to keep this PR from being merged.

Reply:
Thanks, for the thorough review. This kind of feedback is exactly what I am hoping for :) I changed the if statements, to take the model value, if tuning that value is not requested.

Reply:
Tried adding a separate unit tests to check that only the values one desires are actually tuned.

Reply:
That looks reasonable, thanks.